### ISSUE TITLE
Incompatibility with protocol when using sub-protocols

### ISSUE THREAD

**Issue by Finistere** [microsoft/pyright#3396]
**Describe the bug**
Pyright reports an error when using sub-protocols, having issues assigning type variables correctly? Or with type variance? It's unclear to me why it fails, especially the first error where it mixes two type variables: `Value` and `Arg`.

**To Reproduce**
This is the minimal example I could find that has a similar error:
```python
from typing import Generic, TypeVar

from typing_extensions import Protocol

Arg = TypeVar('Arg', contravariant=True)
Value = TypeVar('Value')


class Base(Protocol[Value]):
    def method(self, default: Value) -> Value:
        ...


class Interface(Base[Value], Protocol[Arg, Value]):
    def another(self, arg: Arg) -> None:
        ...


class Inpl(Generic[Arg, Value]):
    def method(self, default: Value) -> Value:
        return default

    def another(self, arg: Arg) -> None:
        return


def f(arg: Arg, value: Value) -> Interface[Arg, Value]:
    return Inpl[Arg, Value]()  # <- Fails here

```
```
scratch_1.py
  scratch_1.py:28:12 - error: Expression of type "Inpl[Arg@f, Value@f]" cannot be assigned to return type "Interface[Arg@f, Value@f]"
    "Inpl[Arg@f, Value@f]" is incompatible with protocol "Interface[Arg@f, Value@f]"
      TypeVar "Value@Base" is invariant
        Type "Value@f" cannot be assigned to type "Arg@f" (reportGeneralTypeIssues)

```

The error can be solved by avoiding sub-protocols:
```python
from __future__ import annotations

from typing import Generic, TypeVar

from typing_extensions import Protocol

Arg = TypeVar('Arg', contravariant=True)
Value = TypeVar('Value')


class Interface(Protocol[Arg, Value]):
    def method(self, default: Value) -> Value:
        ...

    def another(self, arg: Arg) -> None:
        ...


class Inpl(Generic[Arg, Value]):
    def method(self, default: Value) -> Value:
        return  default

    def another(self, arg: Arg) -> None:
        return


def f(arg: Arg, value: Value) -> Interface[Arg, Value]:
    return Inpl[Arg, Value]()
```

While the following looks almost the same, the error isn't exactly:
```python
from typing import Generic, TypeVar

from typing_extensions import Protocol

Arg = TypeVar('Arg', contravariant=True)
Value = TypeVar('Value')


class Base(Protocol[Arg, Value]):
    def method(self, arg: Arg, default: Value) -> Value:
        ...


class Interface(Base[Arg, Value], Protocol[Arg, Value]):
    def another(self) -> Base[Arg, object]:
        ...


class Impl(Generic[Arg, Value]):
    def method(self, arg: Arg, default: Value) -> Value:
        return default

    def another(self) -> Base[Arg, object]:
        return # type: ignore


def f(arg: Arg, value: Value) -> Interface[Arg, Value]:
    return Impl[Arg, Value]()  # <- Fails here

```
```
scratch_1.py
  scratch_1.py:28:12 - error: Expression of type "Impl[Arg@f, Value@f]" cannot be assigned to return type "Interface[Arg@f, Value@f]"
    "Impl[Arg@f, Value@f]" is incompatible with protocol "Interface[Arg@f, Value@f]"
      TypeVar "Value@Interface" is invariant
        Type "Value@Interface" cannot be assigned to type "Value@f" (reportGeneralTypeIssues)

```

Using `Value` instead of explicitly using `object` does work though:

```python
from typing import Generic, TypeVar

from typing_extensions import Protocol

Arg = TypeVar('Arg', contravariant=True)
Value = TypeVar('Value')


class Base(Protocol[Arg, Value]):
    def method(self,  arg: Arg, default: Value) -> Value:
        ...


class Interface(Base[Arg, Value], Protocol[Arg, Value]):
    def another(self) -> Base[Arg, Value]:
        ...


class Impl(Generic[Arg, Value]):
    def method(self,  arg: Arg, default: Value) -> Value:
        return default

    def another(self) -> Base[Arg, Value]:
        return self


def f(arg: Arg, value: Value) -> Interface[Arg, Value]:
    return Impl[Arg, Value]()

```

All examples work with `mypy --strict` version `0.950`, but maybe for the wrong reasons. :)

**Expected behavior**
No errors.

**VS Code extension or command-line**
Command-line, version `1.1.241` on Python `3.9.11`.

**Comment by erictraut** [comment]
There are two issues here. The first one is that the pyright is not evaluating the intended order of type parameters for class `Interface`. By default, type parameters are assigned in the order in which unique TypeVars appear, which means that in the first code sample, `Interface` would have two type parameters in the order `Value` and `Arg`. The code sample assumes the opposite ordering.

The documentation in PEP 544 is not explicit about whether `Protocol[...]` (where `...` is a comma-delimited list of TypeVars) is meant to imply the ordering of type parameters independent of the order in which they appear. It appears that mypy makes this assumption, which is a reasonable interpretation. I've updated the logic in pyright accordingly, so it now honors the order of type parameters as specified in the `Protocol[...]` subexpression.

I've also added an additional check for the case where `Protocol[...]` and `Generic[...]` are both used as base classes. This condition is flagged as an error by mypy even though it doesn't generate a runtime exception. Since this can lead to ambiguities, I think it's reasonable to flag it as an error by a type checker.

Even with the above changes, there is still another bug that prevents this code sample from type checking correctly. It is specific to the case where a generic protocol class derives from another protocol class and at least one of the generic type parameters does not appear in any of the methods or attributes declared within the subclass. This is the case in the initial code sample above where `Value` isn't used within the `Interface` class definition. This bug will require additional investigation and diagnosis to fix.

**Comment by erictraut** [comment]
This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.243, which I just published. It will also be included in the next release of pylance.

**Comment by Finistere** [comment]
Late response but I suppose Mypy is applying the same logic for `Protocol` than for `Generic` explained at https://mypy.readthedocs.io/en/stable/generics.html#defining-sub-classes-of-generic-classes

> If Generic[...] is present, then the order of variables is always determined by their order in Generic[...].

Thanks for the patch and the explanation! :)
